Making TransitioningAttributeStorage "non-eager", throttling et al.

cstamas opened the pull request:
Attribute upgrade from now on happens only on attribute STORE operation,
while attribute READ operation just _reads_ them (from new or falls back to old
storage).

Just for review, not for vote yet. This change alone would not be sufficient for sure, coz some tests expect the "old" behavior.

cstamas committed:
Making TransitioningAttributeStorage "non-eager".

Attribute upgrade from now on happens only on attribute STORE operation,
while attribute READ operation just _reads_ them (from new or falls back to old
storage).

velo commented:
+1

cstamas committed:
Implemented Walk throttling.

Walker component got "throttling" support in very non-intrusive way.
Everything works as before, but there is now possibility to implement
real "throttling" too, ie. intentionally slowing down the walks if needed.

This is currently used in AttributeUpgrade background thread only, as we detected
that thread priority has nothing to do with IO (but only with CPU) and that
presence of this thread might still saturate IO of a server (and leaving CPU
under utilized, hence, JVM does all "good" with this low prio thread).

But with this change, every "walk" is now capable of being throttled even without
(or with slight) code change.

The current implementation should be used as "last resort"! If we need this at all,
using the system property override (to set "throttle time in millis") could be used
to empirically "tune" and find best values for given (hence, similar) Nexus instance.

By default, upgrade thread is NOT throttled.

bdemers commented:
We should have tests for this
(not sure the best way to test throttling)
other then that it looks good

cstamas committed:
Improvements to throttle support, and more.

Throttle support has a major improvements, now we can control throughput in
"transaction per second" manner too. Heavily using it in very improved UpgraderThread
that dynamically adapts itself to the limit "TPS".

peterlynch commented on nexus/nexus-api/src/main/java/org/sonatype/nexus/proxy/walker/WalkerThrottleController.java:
```diff
@@ -0,0 +1,77 @@
+
+/**
+ * Interface to controle throttling (if desired) in executing of Walks.
+ * 
```
sp.

peterlynch commented on nexus/nexus-proxy/src/main/java/org/sonatype/nexus/proxy/attributes/upgrade/DefaultAttributeUpgrader.java:
```diff
@@ -0,0 +1,204 @@
+                // nag the user to remove the directory
+                getLogger().info(
+                    "Legacy attribute directory present, but is marked already as upgraded. Please delete, move or rename the \""
+                        + getLegacyAttributesDirectory().getAbsolutePath() + "\" folder." );
```
slf4j logger? could use {} templating

peterlynch commented on nexus/nexus-proxy/src/main/java/org/sonatype/nexus/proxy/attributes/upgrade/DefaultAttributeUpgrader.java:
```diff
@@ -0,0 +1,204 @@
+        }
+        catch ( IOException e )
+        {
+            // hum?
```
finish that thought?

peterlynch commented on nexus/nexus-proxy/src/main/java/org/sonatype/nexus/proxy/attributes/upgrade/UpgraderThread.java:
```diff
@@ -0,0 +1,175 @@
+        }
+        catch ( InterruptedException e )
+        {
+            return;
```
Shouldn't Thread.currentThread().interrupt(); be called here?

peterlynch commented on nexus/nexus-proxy/src/main/java/org/sonatype/nexus/proxy/attributes/upgrade/UpgraderThread.java:
```diff
@@ -0,0 +1,175 @@
+    {
+        try
+        {
+            Thread.sleep( 5000 );
```
nit: personally prefer to see

```
TimeUnit.SECONDS.sleep(5);
```

cstamas commented:
Am closing this pull request for now, to separate it into two pulls, one about substantial fixes (that should go ASAP into master), and one about the "throttling" where we can continue this debate.
